### PR TITLE
Scrub projection transforms

### LINKED ISSUE #2831: Scrubbing `Projection` transformers
Sub task of #2827
RFF, LpNorm, GcNorm, PCA, Whiten, Normalize.
Projection catalog is bad name, and we probably should toss around this transforms across other catalogs.

### PR DESCRIPTION
**Pull request by wschin**
Fix #2831. 

- Iterations 1-4 only contain renaming.
- Iteration 5 removes `ProjectionCatalog`, extend `NormalizationCatalog` to cover all kinds of normalizers, and create `KernelCatalog` for RFF transform.
- Iteration 6 syncs the parameters in PCA and VectorWhiten.

Transforms touched:
RFF, LpNorm, GcNorm, PCA, Whiten, Normalize.

### CODE REVIEW

File: src/Microsoft.ML.PCA/PcaTrainer.cs [hunk 1]
```diff
@@ -90,20 +90,20 @@ internal static class Defaults
...
         /// <param name="rank">The number of components in the PCA.</param>
         /// <param name="oversampling">Oversampling parameter for randomized PCA training.</param>
-        /// <param name="center">If enabled, data is centered to be zero mean.</param>
+        /// <param name="ensureZeroMean">If enabled, data is centered to be zero mean.</param>
         /// <param name="seed">The seed for random number generation.</param>
         internal RandomizedPcaTrainer(IHostEnvironment env,
-            string features,
-            string weights = null,
+            string featureColumnName,
+            string exampleWeightColumnName = null,
```
Comment (sfilipi) [inline src/Microsoft.ML.PCA/PcaTrainer.cs]: >exampleWeightColumnName [](start = 19, length = 23)

does 'example' need to be in the name? just too long..

File: src/Microsoft.ML.StaticPipe/LpNormalizerStaticExtensions.cs [hunk 2]
```diff
@@ -16,7 +16,7 @@ private sealed class OutPipelineColumn : Vector<float>
         {
             public readonly Vector<float> Input;
 
-            public OutPipelineColumn(Vector<float> input, LpNormalizingEstimatorBase.NormalizerKind normKind, bool subMean)
+            public OutPipelineColumn(Vector<float> input, LpNormalizingEstimatorBase.NormKind normKind, bool subMean)
```
Comment (sfilipi) [inline src/Microsoft.ML.StaticPipe/LpNormalizerStaticExtensions.cs]: >NormKind [](start = 85, length = 8)

this feels backwards, from longer, more descriptive name to abbreviation. 

NormalizationKind ?

File: src/Microsoft.ML.StaticPipe/TransformsStatic.cs [hunk 3]
```diff
@@ -1607,16 +1607,16 @@ private sealed class Reconciler : EstimatorReconciler
         /// speciﬁed shift-invariant kernel. With this transform, we are able to use linear methods (which are scalable) to approximate more complex kernel SVM models.
         /// </summary>
         /// <param name="input">The column to apply Random Fourier transfomration.</param>
-        /// <param name="newDim">Expected size of new vector.</param>
-        /// <param name="useSin">Create two features for every random Fourier frequency? (one for cos and one for sin) </param>
+        /// <param name="dimension">The number of random Fourier features to create.</param>
+        /// <param name="useCosAndSinBases">If <see langword="true"/>, use both of cos and sin basis functions to create two features for every random Fourier frequency.  /// Otherwise, only cos bases would be used.</param>
```
Comment (sfilipi) [inline src/Microsoft.ML.StaticPipe/TransformsStatic.cs]: >/// Otherwise, only cos bases would be used.</param> [](start = 171, length = 52)

new line

File: src/Microsoft.ML.PCA/PCACatalog.cs [hunk 4]
```diff
@@ -18,18 +18,18 @@ public static class PcaCatalog
         /// <param name="exampleWeightColumnName">The name of the example weight column (optional).</param>
         /// <param name="rank">The number of principal components.</param>
```
[the remark this answers is not in the document]
Comment (wschin) [reply src/Microsoft.ML.PCA/PCACatalog.cs]: Done.

---
In reply to: [262791198](https://github.com/dotnet/machinelearning/pull/2865#discussion_r262791198) [](ancestors = 262791198)

File: src/Microsoft.ML.PCA/PcaTrainer.cs (hunk 1 above)
Comment (wschin) [reply src/Microsoft.ML.PCA/PcaTrainer.cs]: Yes but it's used everywhere.

---
In reply to: [262785707](https://github.com/dotnet/machinelearning/pull/2865#discussion_r262785707) [](ancestors = 262785707)

**Comment by artidoro** [comment]
I know @TomFinley  has some opinions about how MLContext should be organized.
In your PR you remove the `Transforms.Projection` catalog section and instead create three others:

- `Transforms.DimensionReduction`
  - VectorWhitening
  - PCA
- `Transforms.KernelExpansion`
  - CreateRandomFourierFeatures
- `Transforms.Normalization`
  - Normalize
  - LpNormalize
  - GlobalContrastNormalize

Since this introduces quite a few sections I think @TomFinley  should take a look before we can check it in.

File: src/Microsoft.ML.Transforms/KernelCatalog.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    public static class KernelExpansionCatalog
+    {
+        /// <summary>
+        /// Takes column filled with a vector of floats and maps its to a random low-dimensional feature space.
+        /// </summary>
+        /// <param name="catalog">The transform's catalog.</param>
+        /// <param name="outputColumnName">Name of the column resulting from the transformation of <paramref name="inputColumnName"/>.</param>
+        /// <param name="inputColumnName">Name of column to transform. If set to <see langword="null"/>, the value of the <paramref name="outputColumnName"/> will be used as source.</param>
+        /// <param name="dimension">The number of random Fourier features to create.</param>
+        /// <param name="useCosAndSinBases">If <see langword="true"/>, use both of cos and sin basis functions to create two features for every random Fourier frequency.  /// Otherwise, only cos bases would be used.</param>
```
Comment (artidoro) [inline src/Microsoft.ML.Transforms/KernelCatalog.cs]: >/// [](start = 171, length = 3)

new line

File: src/Microsoft.ML.Transforms/KernelCatalog.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        /// <example>
+        /// <format type="text/markdown">
+        /// <![CDATA[
+        /// [!code-csharp[CreateRandomFourierFeatures](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/ProjectionTransforms.cs?range=1-6,12-112)]
+        /// ]]>
+        /// </format>
+        /// </example>
+        public static RandomFourierExpansionEstimator RandomFourierExpand(this TransformsCatalog.KernelExpansionTransforms catalog,
```
Comment (artidoro) [inline src/Microsoft.ML.Transforms/KernelCatalog.cs]: >RandomFourierExpansionEstimator [](start = 22, length = 31)

Why Random? I like the Expansion part, but why Random? I don't think the process it that random right?

**Comment by TomFinley** [comment]
> Since this introduces quite a few sections I think @TomFinley should take a look before we can check it in.

Well, I feel like this is introducing *too* much categorization @wschin and @artidoro. So if the categories start to have one or two items, then I wonder what the point of categorization is. So why not just not categorize if we get to that point? Does anything really bad happen if transforms isn't endlessly categorized? This is starting to resemble a Linnaeusian taxonomy.

I see a few end states with categorization. We have in this PR at the time I am writing descriptive categories, but they are too fine grained. We had previously categories that were, while accurate descriptions, utterly useless in terms of communicating anything to users (projection). The alternative is that we don't categorize these at all, and so I wonder if that's the most desirable state if we can't come up with good *descriptive* categories that don't suffer from having only one or two items in them.

So how bad is it if we just put these things in transforms?

File: src/Microsoft.ML.Transforms/KernelCatalog.cs (hunk 6 above)
Comment (wschin) [reply src/Microsoft.ML.Transforms/KernelCatalog.cs]: Formal name: randomized Fourier kernel approximation.

---
In reply to: [263199601](https://github.com/dotnet/machinelearning/pull/2865#discussion_r263199601) [](ancestors = 263199601)

File: src/Microsoft.ML.Transforms/KernelCatalog.cs (hunk 5 above)
Comment (wschin) [reply src/Microsoft.ML.Transforms/KernelCatalog.cs]: Nice catch.

---
In reply to: [263198787](https://github.com/dotnet/machinelearning/pull/2865#discussion_r263198787) [](ancestors = 263198787)

**Comment by wschin** [comment]
Not bad! Let's do this.

---
In reply to: [470339262](https://github.com/dotnet/machinelearning/pull/2865#issuecomment-470339262) [](ancestors = 470339262)

File: src/Microsoft.ML.Transforms/NormalizerCatalog.cs [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        /// <example>
+        /// <format type="text/markdown">
+        /// <![CDATA[
+        /// [!code-csharp[Normalize](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/Normalizer.cs)]
+        /// ]]>
+        /// </format>
+        /// </example>
+        public static NormalizingEstimator Normalize(this TransformsCatalog catalog,
+           string outputColumnName, string inputColumnName = null,
+            NormalizingEstimator.NormalizerMode mode = NormalizingEstimator.NormalizerMode.MinMax)
```
Comment (artidoro) [inline src/Microsoft.ML.Transforms/NormalizerCatalog.cs]: >NormalizerMode [](start = 33, length = 14)

How about NormalizationMode?

File: src/Microsoft.ML.Transforms/NormalizerCatalog.cs [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        /// <param name="columns">The pairs of input and output columns.</param>
+        /// <example>
+        /// <format type="text/markdown">
+        /// <![CDATA[
+        /// [!code-csharp[Normalize](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/Normalizer.cs)]
+        /// ]]>
+        /// </format>
+        /// </example>
+        public static NormalizingEstimator Normalize(this TransformsCatalog catalog,
+            NormalizingEstimator.NormalizerMode mode,
```
Comment (artidoro) [inline src/Microsoft.ML.Transforms/NormalizerCatalog.cs]: >NormalizerMode [](start = 33, length = 14)

NormalizationMode?

**Comment by Ivanidzo4ka** [comment]
>             public string WeightColumn = PrincipalComponentAnalysisEstimator.Defaults.WeightColumn;

Can we update it to our default `exampleWeightsColumnName` i think.

---
Refers to: src/Microsoft.ML.PCA/PcaTransformer.cs:44 in d4edc48. [](commit_id = d4edc4889390ce4914fe6246286a6799cca6e100, deletion_comment = False)

**Comment by Ivanidzo4ka** [comment]
>         }

Need better names.
NormKind is not great. StdDev is not great, Linf is not great.

---
Refers to: src/Microsoft.ML.Transforms/GcnTransform.cs:657 in d4edc48. [](commit_id = d4edc4889390ce4914fe6246286a6799cca6e100, deletion_comment = False)

File: docs/samples/Microsoft.ML.Samples/Dynamic/ProjectionTransforms.cs [hunk 9]
```diff
@@ -37,7 +37,7 @@ public static void Example()
             };
 
             // A pipeline to project Features column into Random fourier space.
-            var rffPipeline = ml.Transforms.Projection.CreateRandomFourierFeatures(nameof(SamplesUtils.DatasetUtils.SampleVectorOfNumbersData.Features), newDim: 4);
+            var rffPipeline = ml.Transforms.RandomFourierExpand(nameof(SamplesUtils.DatasetUtils.SampleVectorOfNumbersData.Features), dimension: 4);
```
Comment (Ivanidzo4ka) [inline docs/samples/Microsoft.ML.Samples/Dynamic/ProjectionTransforms.cs]: >Expand [](start = 57, length = 6)

Expand assumes you can only grow in size.
Maybe just `RandomFourierDecomposition`?

Comment (Ivanidzo4ka) [inline docs/samples/Microsoft.ML.Samples/Dynamic/ProjectionTransforms.cs]: >dimension [](start = 134, length = 9)

if it's `rank` in Whitening and PCA, maybe it also should be `rank`?

Comment (Ivanidzo4ka) [reply docs/samples/Microsoft.ML.Samples/Dynamic/ProjectionTransforms.cs]: Or `dimension` in Whitening and PCA

---
In reply to: [263557554](https://github.com/dotnet/machinelearning/pull/2865#discussion_r263557554) [](ancestors = 263557554)

**Comment by Ivanidzo4ka** [comment]
>         Pca,

PrincipalComponentAnalysis
and 
ZeroPhaseComponentAnalysis


---
Refers to: src/Microsoft.ML.Mkl.Components/VectorWhitening.cs:42 in d4edc48. [](commit_id = d4edc4889390ce4914fe6246286a6799cca6e100, deletion_comment = False)

**Comment by Ivanidzo4ka** [comment]
>         public static RandomizedPcaTrainer RandomizedPca(this AnomalyDetectionCatalog.AnomalyDetectionTrainers catalog,

ApproximatePrincipalComponentsAnalysis?

---
Refers to: src/Microsoft.ML.PCA/PCACatalog.cs:56 in d4edc48. [](commit_id = d4edc4889390ce4914fe6246286a6799cca6e100, deletion_comment = False)

**Comment by Ivanidzo4ka** [comment]
>     public sealed class RandomizedPcaTrainer : TrainerEstimatorBase<AnomalyPredictionTransformer<PcaModelParameters>, PcaModelParameters>

can we move away from PCA to `PrincipalComponentAnalysis` ?

---
Refers to: src/Microsoft.ML.PCA/PcaTrainer.cs:42 in d4edc48. [](commit_id = d4edc4889390ce4914fe6246286a6799cca6e100, deletion_comment = False)

File: docs/samples/Microsoft.ML.Samples/Dynamic/ProjectionTransforms.cs (hunk 9 above)
Comment (wschin) [reply docs/samples/Microsoft.ML.Samples/Dynamic/ProjectionTransforms.cs]: Your argument is not correct. This is a process for approximating (infinite) number of features.

Comment (wschin) [reply docs/samples/Microsoft.ML.Samples/Dynamic/ProjectionTransforms.cs]: Let's do rank.

**Comment by wschin** [comment]
>         public static RandomizedPcaTrainer RandomizedPca(this AnomalyDetectionCatalog.AnomalyDetectionTrainers catalog,

Will do AnalyzeRandomizedPrincipleComponents

---
In reply to: [470688500](https://github.com/dotnet/machinelearning/pull/2865#issuecomment-470688500) [](ancestors = 470688500)

---
Refers to: src/Microsoft.ML.PCA/PCACatalog.cs:56 in d4edc48. [](commit_id = d4edc4889390ce4914fe6246286a6799cca6e100, deletion_comment = False)

**Comment by wschin** [comment]
>     public sealed class RandomizedPcaTrainer : TrainerEstimatorBase<AnomalyPredictionTransformer<PcaModelParameters>, PcaModelParameters>

Sure. I will replace `RandomizedPcaTrainer` with `RandomizedPrincipalComponentAnalyzer`.

---
In reply to: [470690330](https://github.com/dotnet/machinelearning/pull/2865#issuecomment-470690330) [](ancestors = 470690330)

---
Refers to: src/Microsoft.ML.PCA/PcaTrainer.cs:42 in d4edc48. [](commit_id = d4edc4889390ce4914fe6246286a6799cca6e100, deletion_comment = False)

**Comment by wschin** [comment]
>         Pca,

Sure.

---
In reply to: [470687736](https://github.com/dotnet/machinelearning/pull/2865#issuecomment-470687736) [](ancestors = 470687736)

---
Refers to: src/Microsoft.ML.Mkl.Components/VectorWhitening.cs:42 in d4edc48. [](commit_id = d4edc4889390ce4914fe6246286a6799cca6e100, deletion_comment = False)

**Comment by wschin** [comment]
>         }

How about
```csharp
        public enum NormFunction : byte
        {
            L2Norm = 0,
            StandardDeviation = 1,
            L1Norm = 2,
            InfinityNorm = 3
        
```

---
In reply to: [470682324](https://github.com/dotnet/machinelearning/pull/2865#issuecomment-470682324) [](ancestors = 470682324)

---
Refers to: src/Microsoft.ML.Transforms/GcnTransform.cs:657 in d4edc48. [](commit_id = d4edc4889390ce4914fe6246286a6799cca6e100, deletion_comment = False)

**Comment by wschin** [comment]
>             public string WeightColumn = PrincipalComponentAnalysisEstimator.Defaults.WeightColumn;

Sure. Btw, it should be "E" and without "s".

---
In reply to: [470681275](https://github.com/dotnet/machinelearning/pull/2865#issuecomment-470681275) [](ancestors = 470681275)

---
Refers to: src/Microsoft.ML.PCA/PcaTransformer.cs:44 in d4edc48. [](commit_id = d4edc4889390ce4914fe6246286a6799cca6e100, deletion_comment = False)

File: src/Microsoft.ML.Transforms/NormalizerCatalog.cs (hunk 8 above)
Comment (wschin) [reply src/Microsoft.ML.Transforms/NormalizerCatalog.cs]: Ok.

---
In reply to: [263234251](https://github.com/dotnet/machinelearning/pull/2865#discussion_r263234251) [](ancestors = 263234251)

File: src/Microsoft.ML.Transforms/NormalizerCatalog.cs (hunk 7 above)
Comment (wschin) [reply src/Microsoft.ML.Transforms/NormalizerCatalog.cs]: Ok.

---
In reply to: [263234205](https://github.com/dotnet/machinelearning/pull/2865#discussion_r263234205) [](ancestors = 263234205)

File: docs/samples/Microsoft.ML.Samples/Dynamic/ProjectionTransforms.cs (hunk 9 above)
Comment (Ivanidzo4ka) [reply docs/samples/Microsoft.ML.Samples/Dynamic/ProjectionTransforms.cs]: [https://www.merriam-webster.com/dictionary/expand](https://www.merriam-webster.com/dictionary/expand)
1: to open up : UNFOLD
2 : to increase the extent, number, volume, or scope of : ENLARGE
3a : to express at length or in greater detail
bla bla bla.

In this example you take vector of 10 features and map it into 4 dimensions.
How this is expand? It's a RandomFourierShrink!
I now you can put bigger number, but would be nice to have verb which serve both cases, rather than one of them.
RandomFourierMapping maybe?

---
In reply to: [263594873](https://github.com/dotnet/machinelearning/pull/2865#discussion_r263594873) [](ancestors = 263594873)

Comment (wschin) [reply docs/samples/Microsoft.ML.Samples/Dynamic/ProjectionTransforms.cs]: Ok. I will do
```csharp
        public static RandomFourierKernelMappingEstimator RandomFourierKernelMap(this TransformsCatalog catalog,

```

---
In reply to: [263609876](https://github.com/dotnet/machinelearning/pull/2865#discussion_r263609876) [](ancestors = 263609876,263594873)
